### ISSUE TITLE
Rename host to hostname

### ISSUE THREAD

**Issue by sholladay** [rethinkdb/rethinkdb#6293]
Making libraries that interact with RethinkDB programmatically, I have to do some awkward and messy dancing around the API due to the incompatibility of `r.connect()`'s option names in relation to most URL parsing / stringifying libraries.

https://rethinkdb.com/api/javascript/connect/

As is, RethinkDB's `host` option explicitly excludes the `port`, which is wrong. Port is optional in a host, but the capability of including a port is specifically what separates a _host_ from a _hostname_.

I would very much appreciate it if RethinkDB and its libraries would conform to the [URL spec](https://url.spec.whatwg.org) and do both of the following:

1. Add a `hostname` option (really just rename `host`).
2. Either give `host` the ability to include a `port` or get rid of it.

**Comment by AtnNn** [comment]
I would gladly accept a pull request that does something like this, but IPv6-aware:

```
if host is not None and ':' in host:
  if port is not None:
    raise Something()
  [host, port] = host.split(':', 1)
```

However I don't see any strong advantage in adding a `hostname`. It breaks DRY and backwards compatibility.

How would a different argument name make your code less awkward and messy?

Note that for backwards compatibility and consistency with other drivers, the `host` and `port` arguments are the first and second positional arguments.

**Comment by sholladay** [comment]
Simple example:

```js
const url = require('url');

url.parse('rethinkdb://localhost:3000');
```
Output:
```js
{
    protocol: 'rethinkdb:',
    slashes: true,
    auth: null,
    host: 'localhost:3000',
    port: '3000',
    hostname: 'localhost',
    hash: null,
    search: null,
    query: null,
    pathname: null,
    path: null,
    href: 'rethinkdb://localhost:3000'
}
```

I cannot pass this object to `r.connect()`. The URL parser follows the spec, RethinkDB does not.

**Comment by AtnNn** [comment]
> I cannot pass this object to r.connect()

Do you mean `r.connect(*url_info)`? That is not a style of call that I would want to support. It isn't backwards compatible and it ignores parts of the url such as the protocol and path.

You can easily do `r.connect(url_info['hostname'], url_info['port'])`. It's longer, but calling it awkward and messy is an over-exaggeration.

With a change similar to the one I suggest above, you could shorten it to `r.connect(url_info['host'])`.

**Comment by sholladay** [comment]
> That is not a style of call that I would want to support

Seems like we're on different pages here. I'm not suggesting changing the call signature whatsoever. I'm suggesting changing the name of a property on the options object.

I just need to be able to _handle user input_ and pass something reasonable to RethinkDB and have RethinkDB do something reasonable with the reasonable thing I gave to it.

As a downstream _library author_, I am expected to do things like this...

```js
let connectOpts = {};
if (typeof input === 'string') {
    connectOpts = url.parse(input);
}
if (input && typeof input === 'object') {
    Object.assign(connectOpts, input);
}
r.connect(connectOpts);
```

I am over-simplifying things here so as to not confuse the issue. The point is:

 - There is a standard for how to represent URLs as objects.
 - RethinkDB is violating the semantics of that standard.
 - It is currently necessary to massage data in various places.
 - The massaging that is necessary is destructive to the object.
 - These objects often don't belong to us.
 - Programmers are bad at mutating objects.

In other words, it genuinely opens up a small can of worms. I don't mean to exaggerate. It's just something that I think should be tucked into the next major version.

**Comment by AtnNn** [comment]
I think I understand you very well. You want to be able to pass the result of things like `url.parse` directly to `r.connect`, but it uses a different naming convention.

> * There is a standard for how to represent URLs as objects.

* No, it's a convention. Standards such as https://url.spec.whatwg.org/ defines both `host` and `hostname` as identical, and they exclude the port.

* `r.connect` does not expect a url or url object, and one shouldn't be passed to it. URL fields such as `protocol`, `path` and `query` have no meaning to `r.connect`, and the semantics of the URL `auth` field conflict with that of the `r.connect` `auth` option, which is deprecated.

>  * The massaging that is necessary is destructive to the object.

```
info = url.parse(input);
connectOpts = { host: info.hostname, port: info.port };
```

It's not a small can of worms, it's an extra line or two. And it's necessary if you want your library to be forward-compatible.

**Comment by sholladay** [comment]
> Standards such as https://url.spec.whatwg.org/ defines both host and hostname as identical, and they exclude the port.

Nah, read closely, "host attribute’s getter does return a URL-port string".

The semantics of these things are so important that Node core is even supporting two URL APIs simultaneously in order to achieve spec compliance without breaking the web. See the new [WHATWG API](https://nodejs.org/api/url.html#url_the_whatwg_url_api). Both APIs have the semantics I describe. So does [window.URL](https://developer.mozilla.org/en-US/docs/Web/API/Window/URL) in browsers, not to mention the much more ancient [location properties](http://stackoverflow.com/questions/6725890/location-host-vs-location-hostname-and-cross-browser-compatibility). This is a thing, I promise. 😉 

> r.connect does not expect a url or url object, and one shouldn't be passed to it

I'm not asking for it to make use of URL objects in their entirety or expecting otherwise. I am only asking for it to be compatible with the widely accepted definitions of the properties it does use in cases where it makes sense.

> auth field conflict with that of the r.connect auth option, which is deprecated

Well perfect, so there is already a pending breaking change to the schema. When `auth` is removed, this is a perfect opportunity.

> It's not a small can of worms, it's an extra line or two

That's incorrect. I could post more code to prove it if really necessary. Suffice it to say the optionality of URL objects makes processing them properly take more than two lines.

**Comment by srh** [comment]
> I cannot pass this object to r.connect().

Of course not!  Why would you pass a whole object from one API to the optargs object of a completely unrelated one?  That's absolutely crazy!  What if RethinkDB adds new options named "protocol" or "auth"?  What you should do is pass the host: and port: options explicitly.

It turns out that the fact that RethinkDB's names conflict with what you want ends up being a feature, it stops you from shooting yourself in the foot with bad engineering.

**Comment by AtnNn** [comment]
> I'm not asking for it to make use of URL objects in their entirety or expecting otherwise. I am only asking for it to be compatible with the widely accepted definitions of the properties it does use in cases where it makes sense.

We are compatible with the two fields of URL objects where it makes sense: the `hostname` and the `port`. However we call them `host` and `port`. These names have not been a problem before and ensure backwards compatibility.

We are not going to add support for using `r.connect` in the way you tried to use it in your code examples. I've shown how adding a single non-destructive line of code lets you use this API as-is.

**Comment by srh** [comment]
Closing -- the name could have been better but stability is more important.
